homeassistant/components/life360/__init__.py

```diff
@@ -28,184 +22,174 @@
-                cv.time_period, cv.positive_timedelta
-            ),
-            vol.Optional(CONF_MEMBERS): _EXCL_INCL_SLUG_LIST,
-            vol.Optional(CONF_PREFIX, default=DEFAULT_PREFIX): vol.All(
-                vol.Any(None, cv.string), _prefix
-            ),
-            vol.Optional(
-                CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
-            ): cv.time_period,
-            vol.Optional(CONF_SHOW_AS_STATE, default=[]): vol.All(
-                cv.ensure_list, [vol.In(SHOW_AS_STATE_OPTS)]
-            ),
-            vol.Optional(CONF_WARNING_THRESHOLD): _THRESHOLD,
-        }
-    ),
-    _thresholds,
-)
+    return True
 
-CONFIG_SCHEMA = vol.Schema({DOMAIN: LIFE360_SCHEMA}, extra=vol.ALLOW_EXTRA)
 
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Migrate config entry."""
+    LOGGER.debug("Migrating config entry from version %s", entry.version)
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up integration."""
-    conf = config.get(DOMAIN, LIFE360_SCHEMA({}))
-    hass.data[DOMAIN] = {"config": conf, "apis": {}}
-    discovery.load_platform(hass, Platform.DEVICE_TRACKER, DOMAIN, None, config)
-
-    if CONF_ACCOUNTS not in conf:
-        return True
-
-    # Check existing config entries. For any that correspond to an entry in
-    # configuration.yaml, and whose password has not changed, nothing needs to
-    # be done with that config entry or that account from configuration.yaml.
-    # But if the config entry was created by import and the account no longer
-    # exists in configuration.yaml, or if the password has changed, then delete
-    # that out-of-date config entry.
-    already_configured = []
-    for entry in hass.config_entries.async_entries(DOMAIN):
-        # Find corresponding configuration.yaml entry and its password.
-        password = None
-        for account in conf[CONF_ACCOUNTS]:
-            if account[CONF_USERNAME] == entry.data[CONF_USERNAME]:
-                password = account[CONF_PASSWORD]
-        if password == entry.data[CONF_PASSWORD]:
-            already_configured.append(entry.data[CONF_USERNAME])
-            continue
-        if (
-            not password
-            and entry.source == config_entries.SOURCE_IMPORT
-            or password
-            and password != entry.data[CONF_PASSWORD]
-        ):
-            hass.async_create_task(hass.config_entries.async_remove(entry.entry_id))
-
-    # Create config entries for accounts listed in configuration.
-    for account in conf[CONF_ACCOUNTS]:
-        if account[CONF_USERNAME] not in already_configured:
-            hass.async_create_task(
-                hass.config_entries.flow.async_init(
-                    DOMAIN,
-                    context={"source": config_entries.SOURCE_IMPORT},
-                    data=account,
-                )
-            )
+    if entry.version == 1:
+        entry.version = 2
```

Comment:
We don't need to migrate version if we just need to add data to the config entry. We can update the config entry during config entry setup if the config entry is missing some data.

Reply:
The problem is these options did not used to be stored in the config entry, they came from configuration.yaml. This was all originally implemented before option flows were created/possible. So, we need to get the options from yaml and add them to the config entry. How will we know if that has or has not been done if the version doesn't change? I'm not sure what you're asking for is possible.

Reply:
So, my current implementation only adds an option to the config entry options dict if its value is non-default. But I suppose I could add all the options, and just set their values to their default if not set by the user. Then I could do like you say -- i.e., keep the version # the same and just add any missing option. That should work.